### PR TITLE
gh-105927: Add PyWeakref_GetRef() function

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-105927

----
:books: Documentation preview :books:: https://cpython-previews--105932.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
Once this PR will be merged, I will create a follow PR to replace PyWeakref_GET_OBJECT() with PyWeakref_GetRef(). I have it locally, but GitHub doesn't let me create a "patch serie".

File: Objects/weakrefobject.c
Comment (Eclips4) [inline Objects/weakrefobject.c]:
```suggestion
    assert(PyWeakref_Check(ref));
```

**Comment by encukou** [comment]
Please don't return NULL without an exception set.

**Comment by vstinner** [comment]
It doesn't return NULL with an exception set.

**Comment by vstinner** [comment]
I changed the API to:

```
int PyWeakref_GetRef(PyObject *ref, PyObject **pobj)
```

I also added tests.

@erlend-aasland @encukou: Please review the updated PR.

File: Doc/c-api/weakref.rst
Comment (erlend-aasland) [inline Doc/c-api/weakref.rst]:
```suggestion
   Get the referenced object from a weak reference, *ref*, into *\*pobj*.
```

File: Doc/c-api/weakref.rst
Comment (erlend-aasland) [inline Doc/c-api/weakref.rst]:
```suggestion
   If the referent is no longer live, set *\*pobj* to ``NULL`` and return 0.
```

File: Objects/weakrefobject.c
Comment (erlend-aasland) [inline Objects/weakrefobject.c]: I think this behaviour change should get it its own PR and NEWS item.

**Comment by erlend-aasland** [comment]
Using an `int` return aligns error checking with the well-known C idiom of `if (fn() < 0) { /* err handler */`; IMO, that makes for a better API. Thanks!

File: Objects/weakrefobject.c
Comment (vstinner) [reply Objects/weakrefobject.c]: I thought that it would be nice to make the two functions consistent, but I'm not motivated to write a dedicated PR just for that. I prefer to restrict this PR to PyWeakref_GetRef(), so I revert it.

**Comment by vstinner** [comment]
@erlend-aasland: I addressed your review.

File: Objects/weakrefobject.c
Comment (erlend-aasland) [reply Objects/weakrefobject.c]: Sounds good!

**Comment by erlend-aasland** [comment]
> @erlend-aasland: I addressed your review.

Thanks! Looks good.

**Review by erlend-aasland — APPROVED** [review]
LGTM; I'm interested in Petr's review, though :)

**Comment by erlend-aasland** [comment]
2de389766342ee18907130debf63a0336b0bbb45 is in itself a nice docs update that could be backported through 3.11! :)

File: Doc/c-api/weakref.rst
Comment (encukou) [inline Doc/c-api/weakref.rst:58]: Let's consider returning 1 if `**pobj` was set to an object, and 0 if it's set to NULL: distinguishing “missing result” from “full sucess” in the return value miiiiight be better API in general (especially for a `PyIter_NextItem`), and maaaaaybe it'd be worth the extra check to be consistent.
I [posted about this elsewhere](https://github.com/python/cpython/issues/105201#issuecomment-1600491761), let's not block the PR on it. There's enough time to change it in 3.13.

**Review by encukou — APPROVED** [review]
Thank you! This looks good.

**Comment by vstinner** [comment]
I dislike "3 states" return value: error, case 1 or case 2. It forces to store the return value in a variable and uses 2 if to cover the 3 code paths.

Also, IMO the API is more error prone since someone may write the wrong test for the error case.

Well, I don't know how much is just m personal taste and what is objective here 😁

I prefer to only return -1 on error, and 0 for the other cases. So I don't need an extra variable.

File: Doc/c-api/weakref.rst
Comment (vstinner) [reply Doc/c-api/weakref.rst:58]: Since the devguide got updated to include this new guideline, I created https://github.com/python/cpython/pull/106561 to return 1 on success.

**Pull request merged**
